### PR TITLE
Stabilize `underscore_imports`

### LINKED ISSUE #48216: Tracking issue for RFC #2166: impl-only-use
This is a tracking issue for the RFC "impl-only-use " (rust-lang/rfcs#2166).

**Steps:**

- [x] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?)
- [x] Adjust documentation ([see instructions on forge][doc-guide])
- [x] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

### PR DESCRIPTION
**Pull request by petrochenkov**
Closes https://github.com/rust-lang/rust/issues/48216

### CODE REVIEW

**Comment by petrochenkov** [comment]
## Stabilization report

Feature name: `#![feature(underscore_imports)]`
Version target: 1.32 (2019-01-18)

Implemented in https://github.com/rust-lang/rust/pull/48922 and landed 8.5 months ago.
Tracking issue is https://github.com/rust-lang/rust/issues/48216.
Documentation issue is https://github.com/rust-lang-nursery/reference/issues/470.
Tests for the feature can be found in `src/test/ui/rfc-2166-underscore-imports`.

### What is being stabilized

`use path as _;` imports the path `path` and renames it into some unique identifier that can't be written explicitly.
For traits this means `Trait`'s methods are now in scope in module/block into which `use Trait as _;` is placed.
For non-traits the import has no effects and is reported as unused by the `unused_imports` lint.

Additional effect on 2018 edition is that `use my_crate::a::b as _;` can refer to an extern crate `my_crate` and link it.
However, the import is still reported as unused in this case, unless it's of the form `use my_crate as _;` (add test).

`extern crate name as _;` imports the crate `name` and renames it into some unique indentifier that can't be written explicitly.
The crate imported this way is linked, so the feature can be used for link-only crates, and the crate is not reported as unused in this case (add test).
If `#[macro_use]` is attached to such crate it works as usual and imports macros from `name` into prelude (add test?).

### Some details and corner cases

Despite the name introduced with `_` being unnameable explicitly, it still can be fetched by glob imports.
So, something like `use super::*;` reimports traits brought in scope with `use Trait as _;` in the parent module, and their methods become available in the child module.

Underscore imports can be public - `pub use Trait as _;`, this also can make sense in combination with glob imports.

### Open questions

If procedural macro clone the `_` token from its input, will it refer to the same unique identifier?
Right now the answer is yes, the "unique identifier" is created during parsing, so procedural macros (and declarative macros as well, if you try hard enough (add test)) can clone it while keeping its identity.
It's possible to implement it differently, so unique identifiers are only created during name resolution, after macro expansion.

### Known issues

"Unique identifiers" (gensyms) are lost in crate metadata, so all `_`s became same after being imported from another crate.
This is a hard issue to solve, but it doesn't prevent the common use case of `_` imports for importing tratis, which doesn't normally assume reexporting `_` imports to other crates.

`_` imports and other imports in general, are known to be buggy when used together with macros 2.0 (`macro` items or `Span::def_site`), but this only affects unstable code.

**Comment by petrochenkov** [comment]
I'm especially interested in opinions regarding interactions with proc macros ("Open questions" in https://github.com/rust-lang/rust/pull/56303#issuecomment-442274118).
cc @dtolnay

**Review by Centril — APPROVED** [review]
Looks good pending tests passing and pending some of the additional tests discussed above.

**Comment by Centril** [comment]
Thanks for the thoroughly detailed report <3. 

I propose that we stabilize `#![feature(underscore_imports)]` with a version target of 1.32 (or later if necessary...) for all editions (2015 and 2018). As for the open questions re. proc macros, the current behavior seems correct to me, but it seems a fairly small thing that can be tweaked during this review process.

@rfcbot merge

**Comment by dtolnay** [comment]
To make sure I understand the procedural macro consideration: would the following be an example of the edge case you have in mind?

```rust
#[my_macro]
use path::to::Trait as _;

// expands to

use path::to::Trait1 as _;
use path::to::Trait2 as _; // the same underscore token
```

and these would be conflicting imports because they reuse the same gensym?

**Comment by petrochenkov** [comment]
@dtolnay 
Yes, exactly.

The declarative macro equivalent is
```rust
#![feature(underscore_imports)]
#![allow(unused)]

macro_rules! m {
    ($item: item) => { $item $item }
}

m!(use std as _;); // ERROR the name `_` is defined multiple times

fn main() {}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=a56fe430e06fcb300afcb01a64bcfe8b

**Comment by Centril** [comment]
@dtolnay 

> like how if a macro expands to `let x: Result<_, _> = ...` with the two underscores being the same token, they still operate independently and wouldn't somehow end up being constrained to be the same inferred type.

I agree with this rationale and have amended the proposal accordingly.

**Comment by Nemo157** [comment]
> "Unique identifiers" (gensyms) are lost in crate metadata, so all `_`s become same after being imported from another crate.
> This is a hard issue to solve, but it doesn't prevent the common use case of `_` imports for importing tratis, which doesn't normally assume reexporting `_` imports to other crates.

Re-exporting `_` traits from a prelude is one of the usecases I was looking forward to. [`embedded_hal::prelude`](https://docs.rs/embedded-hal/0.2.2/embedded_hal/prelude/index.html) uses manually renamed traits to try and avoid conflicts, [trying to replace this with `underscore_imports`](https://github.com/Nemo157/embedded-hal/commit/9b02eb69211ecad07b98c78a6148f685b70a1753) with compiles the crate fine, but gives an error related to this when glob imported:

```rust
error[E0428]: the name `_` is defined multiple times
  |
  = note: `_` must be defined only once in the type namespace of this module
```

It would be nice if there were a specific diagnostic for this usecase (and an issue to track implementing a fix?).

**Comment by petrochenkov** [comment]
One more example showing that gensym creation at parse-time is at least suspicious:

```rust
duplicate! {
    // The macro argument is parsed as an unstructured token stream, `_` is parsed as a plain `_` token, gensym is not created.
    // Duplication results in two plain `_` tokens.
    // Two plain `_`s are parsed after the macro is expanded, two different gensyms are created, no name conflict happens during resolution.
    use foo as _;
}

// The macro argument is parsed as an item, gensym is created when parsing `_`.
// Duplication results in two identical gensyms.
// Two gensyms are parsed after the macro is expanded, name conflict happens during resolution.
#[duplicate]
use foo as _;
```

**Comment by petrochenkov** [comment]
All the suggestions and additional tests are implemented in https://github.com/rust-lang/rust/pull/56392.

**Comment by scottmcm** [comment]
@rfcbot concern waiting-on-56392-fix

It looks like there's currently one test in https://github.com/rust-lang/rust/tree/master/src/test/ui/rfc-2166-underscore-imports.  Is that sufficient?  Are there success stories of people trying this out and it working, so people wanting it stabilized to be able to check in?  (I didn't see any in https://github.com/rust-lang/rust/issues/48216, and only a failure story here.)

**Comment by petrochenkov** [comment]
>It looks like there's currently one test in https://github.com/rust-lang/rust/tree/master/src/test/ui/rfc-2166-underscore-imports. Is that sufficient?

Well, this is a small feature. Most of intra-crate cases are basically covered in that test. (And https://github.com/rust-lang/rust/pull/56392 adds some more.)

**Comment by Nemo157** [comment]
Oh, I can give a success story where I've been using this for a while as well 😄 [here](https://github.com/Nemo157/embrio-rs/blob/deefdac474ee2a9767b1fba0f28710596f8804ca/examples/pca10031/examples/hello.rs#L8) I've been using it for bringing in link-only crates (and by the sounds of #56392 that will fix the `allow(unused_imports)`  I had to use).

**Comment by scottmcm** [comment]
@rfcbot resolved waiting-on-56392-fix

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -689,6 +686,8 @@ declare_features! (
     (accepted, extern_crate_item_prelude, "1.31.0", Some(55599), None),
     // Allows use of the :literal macro fragment specifier (RFC 1576)
     (accepted, macro_literal_matcher, "1.31.0", Some(35625), None),
+    // `use path as _;` and `extern crate c as _;`
+    (accepted, underscore_imports, "1.32.0", Some(48216), None),
```
Comment (Centril) [inline src/libsyntax/feature_gate.rs]:
```suggestion
    (accepted, underscore_imports, "1.33.0", Some(48216), None),
```

**Comment by torkleyy** [comment]
Awesome! When exporting traits from a prelude like this, what is `rustdoc` going to generate?

**Comment by petrochenkov** [comment]
@torkleyy 
Some quick testing shows this:

underscore_imports

**Comment by cramertj** [comment]
r=me with @Centril's above fix to the feature gate stabilization version.
